File: 143-reorder-list /main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    // find the middle
    let slow = head;
    let fast = head.next;

    while(fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    // Note: slow is now the middle node

    // second is now the head of the second portion
    let second = slow.next;

    // we don't want to have cycles in the list(cut off the first portion from the second portion)
    slow.next = null;

    // reverse the second portion
    let prev = null

    while (second) {
        const nxt = second.next;
        second.next = prev;
        prev = second;
        second = nxt;
    }

    /* now `prev` has become the head of the second portion which is now reversed. Since we don't want to lose
    the head nodes in general, let's define two new pointers in order traverse the first and second portions(we don't
    want to traverse the lists using head pointers because otherwise we would lose the lists altogether!): */
    let first = head;
    second = prev;

    // merge first and second portions
    while(second) {
        const tmp1 = first.next;
        const tmp2 = second.next;

        first.next = second;
        second.next = tmp1;
        first = tmp1;
        second = tmp2;
    }
};

////////////////////////////////

// using arrays(space complexity O(n))
var reorderList2 = function(head) {
    const nodes = [];

    // Store all nodes from LinkedList into array
    let node = head;
    while (node) {
        nodes.push(node);
        node = node.next;
    }

    // Connecting nodes
    node = nodes[0];
    let counter = 0;
    let length = nodes.length;

    while (counter < length) {
        if (counter === length - 1) {
            node.next = null;
            break;
        }

        if (counter % 2 === 0) {
            node.next = nodes[length - Math.floor(counter / 2) - 1];
        } else {
            node.next = nodes[Math.floor(counter / 2) + 1];
        }

        node = node.next;
        counter++;
    }
};

File: 143-reorder-list /main.py
class ListNode:
	def __int__(self, val=0, next=None):
		self.val = val
		self.next = next


class Solution:
	def reorderList(self, head: ListNode) -> None:
		"""
		Do not return anything, modify head in-place instead.
		"""

		# First, find the middle point of the list, so we need two pointers, one slow and one fast
		slow, fast = head, head.next

		# while fast is not null and fast has not reached the end of the list:
		while fast and fast.next:
			slow = slow.next
			fast = fast.next.next

		# at this point, we have the beginning of the second half of the list which is gonna be the `slow.next`.
		second = slow.next

		# now we have to reverse the second half of the list, but before doing that we need to set slow.next to None, because
		# that slow pointer is gonna become the end of the reordered list(we're splitting the portions).
		slow.next = None

		# at the beginning prev is None and on each iteration, it will be the prev node that we visited so we still have a reference to it in order to
		# reverse the list
		prev = None

		# reverse the second portion of the linked list
		while second:
			tmp = second.next
			second.next = prev
			prev = second
			second = tmp

		# at this point, the second half is reversed, now we have to merge the two halfs of the list
		# Note: The beginning of the second list is gonna be prev. Because after the previous loop was done, second is gonna be
		# set to None, but `prev` is going to be set to whatever the last node that we looked at was and whatever that last node was, is actually
		# gonna be the new head of the second half of the list. In other words, after the loop, as we move forward and reverse the nodes,
		# prev will be the new head of the second reversed list.

		# merge two halfs
		first, second = head, prev

		# we're gonna keep going until one of the pointers is null, so condition would be(first and second), but we also know that the
		# second half of the list could be shorter than the first half(first half would never be shorter than the second half), so we just
		# use `second` as the condition:
		while second:
			# store next nodes of `first` and `second` into temporary variables because we know we're gonna breaking those links between the nodes
			tmp1, tmp2 = first.next, second.next

			first.next = second
			second.next = tmp1

			first = tmp1
			second = tmp2
